### PR TITLE
ReplicatedPG::maybe_handle_cache: do not promote before checking full

### PR DESCRIPTION
**Pull request by athanatos**
Otherwise, ops with need_promote() can order past ops without it
causing an rgw write sequence like:

writefull 0~0
writefull 0~512kb
write 512kb~512kb

to reorder to

writefull 0~0 -- delayed: full
writefull 0~512kb -- delayed: full
write 512kb~512kb -- promote and execute
writefull 0~0 -- execute
writefull 0~512kb -- execute

Fixes: 13240
Signed-off-by: Samuel Just <sjust@redhat.com>

### CODE REVIEW

**Comment by athanatos** [comment]
Passed rados suite, DNM pending rgw suite.

File: src/osd/ReplicatedPG.cc [hunk 1]
```diff
@@ -1977,20 +1977,22 @@ bool ReplicatedPG::maybe_handle_cache(OpRequestRef op,
...
+  bool can_proxy_read =
+    (get_osdmap()->get_up_osd_features() & CEPH_FEATURE_OSD_PROXY_FEATURES) &&
+    !must_promote &&
+    !op->need_promote();
+  bool can_proxy_write =
+    (get_osdmap()->get_up_osd_features() &
+     CEPH_FEATURE_OSD_PROXY_WRITE_FEATURES) &&
+    !must_promote &&
+    !op->need_promote();
+
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: This makes me nervous... must_promote might be true on a read op (see line 1897), and some paths do a read redirect if !can_proxy_read.  See line 2001.  I think adding !must_promote to the condition on line 2000 would fix it?  And 2081, and 2098...

**Comment by athanatos** [comment]
New version, need to test.

**Comment by liewegas** [comment]
lgtm

separately, i think we need to add || must_promote to the READPROXY and READFORWARD cases

**Pull request merged**